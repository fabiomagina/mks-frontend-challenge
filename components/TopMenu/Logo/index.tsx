import styles from './Logo.module.scss'

const Logo = () => {
	return (
		<section className={styles.logo}>
			<h1 className={styles.first__name}>
				MKS
			</h1>
			<h1 className={styles.second__name}>
				Sistemas
			</h1>

		</section>
	);
};

export default Logo;