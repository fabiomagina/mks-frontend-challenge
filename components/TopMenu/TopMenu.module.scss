@import 'styles/_variables.scss';

.top_menu {
    height: 101px;
    background-color: $primary;
    color: $on-primary;
    padding: 0 5vw;
    display: flex;
    justify-content: space-between;
}