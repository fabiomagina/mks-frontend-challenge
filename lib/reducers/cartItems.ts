import { createSlice } from "@reduxjs/toolkit";
import Product from "@/interfaces/IProduct";

interface CartItemsState {
    products: Product[]
    qty: { productId: number, qty: number }[]
}


const cartItemsSlice = createSlice({
    name: 'cartItems',
    initialState: { products: [], qty: [] } as CartItemsState,
    reducers: {
        addToCart: (state: CartItemsState, action) => {
            const { id } = action.payload;
            const product = state.products.find(product => product.id === id);
            if (product) {
                state.qty = state.qty.map(item => {
                    if (item.productId === id) {
                        item.qty++;
                    }
                    return item;
                })
            } else {
                state.products.push(action.payload);
                state.qty.push({ productId: id, qty: 1 });
            }

        },
        removeOneFromCart: (state: CartItemsState, action) => {
            const { id } = action.payload;
            const product = state.products.find(product => product.id === id);
            if (product) {
                state.qty = state.qty.map(item => {
                    if (item.productId === id) {
                        if (item.qty > 1)
                            item.qty--;
                    }
                    return item;
                })
            }
        },
        removeFromCart: (state: CartItemsState, action) => {
            const { id } = action.payload;
            const productIndex = state.products.findIndex(product => product.id === id);
            if (productIndex !== -1) {
                state.products.splice(productIndex, 1);
                state.qty = state.qty.filter(item => item.productId !== id);
            }
        },
        clearCart: (state: CartItemsState) => {
            state.products = [];
            state.qty = [];
        }
    }
})

export const { addToCart, removeOneFromCart, removeFromCart, clearCart } = cartItemsSlice.actions;
export default cartItemsSlice.reducer;