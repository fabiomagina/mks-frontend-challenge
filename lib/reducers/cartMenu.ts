import { createSlice } from "@reduxjs/toolkit";

const cartMenuSlice = createSlice({
    name: 'cartMenu',
    initialState: false,
    reducers: {
        toggleCartMenu: state => !state
    }
})
export const { toggleCartMenu } = cartMenuSlice.actions

export default cartMenuSlice.reducer;