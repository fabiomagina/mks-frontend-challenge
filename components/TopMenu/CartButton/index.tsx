"use client"

import Image from "next/image";
import styles from "./CartButton.module.scss";
import { useAppDispatch, useAppSelector } from "@/lib/hooks";
import { toggleCartMenu } from "@/lib/reducers/cartMenu";

const CartButton = () => {
	const dispatch = useAppDispatch();
  const cartItems = useAppSelector(state => state.cartItems);
  const cartCount = cartItems.qty.reduce((acc, item) => acc + item.qty, 0);
  
  return (
    <button className={styles.cart__button} onClick={() => dispatch(toggleCartMenu())}>
      <Image
        src="/cart.svg"
        alt="Cart"
        width={18}
        height={19} />
      <span className={styles.cart__count}>
        {cartCount}
      </span>
    </button>
  );
};

export default CartButton;