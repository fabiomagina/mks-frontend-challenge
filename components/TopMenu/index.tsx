import CartButton from './CartButton';
import Logo from './Logo';
import styles from './TopMenu.module.scss'

const TopMenu = () => {
  return (
    <section className={styles.top_menu}>
      <Logo />
      <CartButton />
    </section>
  );
};

export default TopMenu;