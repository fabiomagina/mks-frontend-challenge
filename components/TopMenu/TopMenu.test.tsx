import { render, screen } from '@testing-library/react';
import TopMenu from '.';
import { Provider } from 'react-redux';
import cartItems, { addToCart, removeFromCart } from '@/lib/reducers/cartItems';
import { configureStore } from '@reduxjs/toolkit';
import cartMenu from '@/lib/reducers/cartMenu';

const mockStore = configureStore({
    reducer: {
        cartItems,
        cartMenu
    }
});
let store: any;

beforeAll(() => {
    store = mockStore;
});

test('verifies Logo in TopMenu', () => {
    render(<Provider store={store}>
            <TopMenu />
        </Provider>
    );
    const firstName = screen.getByText('MKS');
    expect(firstName).toMatchSnapshot();
});

test('add a product to cart, and verifies if cart value is 1', () => {
    const product = { id: 1, name: 'Product', brand: 'brand', description: 'Description', price: '10', photo: 'photo', updatedAt: '2021-09-01' };
    store.dispatch(addToCart(product))
    render(
        <Provider store={store}>
            <TopMenu />
        </Provider>
    );
    expect(screen.getByText('1')).toMatchSnapshot();
});

test('remove a product and verifies if Cart values equals 0', () => {
    const product = { id: 1, name: 'Product', brand: 'brand', description: 'Description', price: '10', photo: 'photo', updatedAt: '2021-09-01' };
    store.dispatch(removeFromCart(product))

    render(
        <Provider store={store}>
            <TopMenu />
        </Provider>
    );
    const cart = screen.getByText('0');
    expect(cart).toMatchSnapshot();
}
);