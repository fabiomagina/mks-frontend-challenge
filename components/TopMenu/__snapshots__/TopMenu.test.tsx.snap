// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add a product to cart, and verifies if cart value is 1 1`] = `
<span
  class="cart__count"
>
  1
</span>
`;

exports[`remove a product and verifies if Cart values equals 0 1`] = `
<span
  class="cart__count"
>
  0
</span>
`;

exports[`verifies Logo in TopMenu 1`] = `
<h1
  class="first__name"
>
  MKS
</h1>
`;
